1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,208 @@
+Currently, PowerShell functions do not have an effective way of using these resources in the correct manner.
+
+This is further complicated by the fact that any terminating error in the pipeline completely bypasses the `end` named block.
+As such, it cannot be relied upon for cleanup / disposal steps.
+
+This includes any time a user includes the pipeline cmdlet `Select-Object` with the `-First` parameter.
+Once the desired number of objects has been reached, the cmdlet throws an internal and deliberately hidden exception to halt the pipeline.
+This is highly effective, but can lead to undesirable circumstances if essential cleanup steps are then skipped as a result.
+
+Compiled cmdlets do not suffer this issue, as they can simply implement `IDisposable` and place the necessary steps in a `Dispose()` method of their own.
+The pipeline processor will correctly handle **cmdlets** that require disposal steps, but currently not **functions**.
+
+## Motivation
+
+As a developer,<br />
+I want to be able to ensure I am not leaving resources or connections open when a user utilizes a function I have written.
+
+As an administrator,<br />
+I want to be able to ensure that any necessary logging or other necessary administrative tasks cannot be skipped accidentally or deliberately by a user invoking a function in a pipeline and using `Select-Object -First X` to bypass logging which would normally take place in the `end` block.
+
+## Specification
+
+A `dispose{}` block is required for these use cases.
+
+`IDisposable` can be implemented by PowerShell classes, but at the time of writing utilizing a PS-based class for a cmdlet is not feasible.
+There are outstanding issues with PowerShell classes that complicate the possibility of creating a cmdlet with them, especially when working with modules.
+
+### Execution
+
+The `dispose{}` block will execute:
+
+1. After `end{}` during normal command operation, and
+2. During command disposal that occurs when a terminating exception is thrown during pipeline operations.
+    - If the dispose block has already been called (as it is after the `end{}` step), it will simply return immediately.
+
+### Behaviour
+
+The `dispose{}` block silently **discards** anything submitted to the success output stream from within it.
+
+Documentation should advise that `dispose{}` is _not intended_ for any standard output and will completely ignore it.
+Additionally, all other PowerShell data streams are accessible from `dispose{}`, meaning there is no issue sending data using any of the following streams from within `dispose{}`:
+
+- Error
+- Warning
+- Verbose
+- Debug
+- Information
+
+#### Error States
+
+Terminating errors can also be thrown from a `dispose{}` block, which will skip the invocation of the rest of that `dispose{}` block, without preventing other `dispose{}` blocks from being executed.
+
+### Implementation Details
```

Comment:
@PowerShell/powershell-committee would like to have specific language on what happens with Ctrl+C.  Our suggestion would be to emit a message to the user such as "Ctrl+C pressed, attempting to stop processing".  And allow `Dispose{}` to execute which may take some time.  A second Ctrl+C would forcibly kill the pipeline.

Reply:
I agree on the point of submitting a message to the user, that makes sense.

I have to disagree on the ability to force-cancel it though. If it's that critical, they always have the option to terminate the process, same as ever. If we implement a "spam Ctrl+C to cancel", I think it's quite likely that users will over time become accustomed to spamming it, terminating the cleanup code as a habit.

That rather defeats the purpose here, and communicates the wrong message about both what should be put in `dispose{]` and how much a script/module author can actually trust `dispose{}` to do what it says on the tin.

Reply:
it's *that* ambiguity as to why we need more language to describe the behavior specifically :)

Reply:
Yeah, and the issue here is that if you can't override the dispose code, it can forever hang your pipeline.

Reply:
Adding a second Ctrl+C support is additive, so we can certainly defer that til later or if it becomes an issue

Reply:
Agreed, Steve. I acknowledge that there are concerns about it, but I'll just put it this way -- if it was going to be an issue, it more or less _already_ is. Authors of cmdlets can already define a cmdlet which implements `IDisposable` (this is a [documented recommendation](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/cmdlet-input-processing-methods?view=powershell-7#post-processing-operations)).

For that reason I'd agree with James, and I think in terms of communicating it to users, it should be documented / communicated as a direct analogue of `IDisposable`, which it effectively is.

Reply:
> Agreed, Steve. I acknowledge that there are concerns about it, but I'll just put it this way -- if it was going to be an issue, it more or less _already_ is. Authors of cmdlets can already define a cmdlet which implements `IDisposable` (this is a [documented recommendation](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/cmdlet-input-processing-methods?view=powershell-7#post-processing-operations)).

A cmdlet author could also just not implement `StopProcessing` and block the pipeline indefinitely in any block.  A function can also call a CLR method and block indefinitely.  It's always been the user's responsibility to avoid this.

Reply:
Updated the RFC to include this feedback. 🙂

Reply:
Feel free to let me know if the added portions to the RFC are lacking anything from this discussion, as well.

Reply:
Was directed here from the issue in the main repo, so I'll share my two cents.

* I agree strongly with @SeeminglyScience that this does not meaningfully increase the chances of blocking given all of the much more common ways one can hang the shell (calling a .NET method that blocks being the most common/obvious).
* If `dispose` is chosen as the keyword for this, then it should actually be semantically the same as implementing `IDisposable` on a cmdlet.  Not kinda thematically similar.  I used the example of try/catch in the issue - explaining "well, but for some kinds of errors, catch doesn't work the way it does in every other language" over and over is no fun.  If you're going to pick something recognizable, great, but make it actually be that recognizable thing.  If it's something different, give it a different name.
* One potentially tricky thing I can imagine is interacting with strict mode.  It's reasonable to assume that the audience for this feature has a fairly high intersection with the audience for strict mode (pedantic people like me).  There's no terse way of checking whether a variable is defined in a strict mode safe way, and it seems like you'd have no idea or control over when you'd be sent to the `dispose` block here.  That's not a reason not to do it, just highlighting it as a potential concern (or maybe I'm overlooking something?)

Reply:
> If dispose is chosen as the keyword for this, then it should actually be semantically the same as implementing IDisposable on a cmdlet. Not kinda thematically similar.

Yep, this is what it's intended to be, and how it behaves in the pull request I submitted before this RFC, unless I missed something. 🙂 

> https://github.com/PowerShell/PowerShell/pull/9900

Feel free to pull that code and give it a look. I probably have some rebasing to do, given that the code was written almost a year ago now, but it's fully functional and will run just fine. 

> One potentially tricky thing I can imagine is interacting with strict mode.

Interesting point regarding strict mode, indeed. I think perhaps the `?.` operator that was recently introduced will alleviate _some_ of that headache, but yeah strict mode has really pretty much always been a second-class citizen in a lot of ways. I can't definitively say I've made it much easier here, but I don't think I've in any way made it any worse either.

That said, I haven't spoken to a lot of folks who can tolerate scripting in strict mode, to be quite honest. They seem to be fairly few and far between. Personally, I'd _like_ to be able to write code in strict mode, but given the myriad issues around it I've elected not to until they get resolved. There's too much pain for far too little gain where that's concerned currently, but that's another issue entirely, really. 🙂

Reply:
To each their own I suppose. My team has about 250 kloc of homegrown PS modules and scripts. Virtually all of it uses strict mode. We (strict mode users) may be few and far between, but we exist :)

Don’t want to derail this thread or anything, and I agree that I don’t think this makes things worse.  We’d certainly adopt this if it gets merged.

Reply:
I do everything in strict mode. Saves me from lots of issues in large scripts.

In re: terse checking for variable existence, what's wrong with "$null = gv vname -ea 0" ??

Reply:
Nothing inherently wrong with it, but it involves a cmdlet call rather than being available directly in the language, so it comes with a bit of additional overhead that you could argue shouldn't be necessary.

Reply:
That’s not all that terse, puts stuff in $Error (which breaks stuff like JEA startup scripts) and depends on aliases, which some people might want to avoid generally (did a module get imported by a script running in -WhatIf mode, then it probably didn’t export its aliases).  `$ExecutionContext.SessionState.PSVariable.Get()` is what I end up with.

By terse, I meant some qualifying character like $?foo, which would get you $foo if it was defined or null when in strict mode (basically disable strict mode for that variable reference only).  That’s not a specific proposal and is not at all thought out, just an example.